lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java

```diff
@@ -0,0 +1,133 @@
+      final Map<String, Boolean> isStreamingCodecMap) {
+
+    this.factory = factory;
+    this.connectionFactoryMap = connectionFactoryMap;
+    this.isStreamingCodecMap = isStreamingCodecMap;
+  }
+
+  /**
+   * Encodes a network service message to bytes.
+   * @param obj a message
+   * @return bytes
+   */
+  @Override
+  public byte[] encode(final DefaultNSMessage obj) {
+    final Codec codec = connectionFactoryMap.get(obj.getConnectionFactoryId()).getCodec();
+    final Boolean isStreamingCodec = isStreamingCodecMap.get(obj.getConnectionFactoryId());
+
+    try (final ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      try (final DataOutputStream daos = new DataOutputStream(baos)) {
+        daos.writeUTF(obj.getConnectionFactoryId());
+        daos.writeUTF(obj.getSrcId().toString());
+        daos.writeUTF(obj.getDestId().toString());
+        daos.writeInt(obj.getData().size());
+
+        if (isStreamingCodec) {
+          for (final Object rec : obj.getData()) {
+            ((StreamingCodec) codec).encodeToStream(rec, daos);
+          }
+        } else {
+          final Iterable dataList = obj.getData();
+          for (final Object message : dataList) {
+            final byte[] bytes = codec.encode(message);
+            daos.writeInt(bytes.length);
+            daos.write(bytes);
+          }
+        }
+        return baos.toByteArray();
+      }
+    } catch (final IOException e) {
+      throw new RuntimeException("IOException", e);
+    }
+  }
+
+  /**
+   * Decodes a network service message from bytes.
+   *
+   * @param data bytes
+   * @return a message
+   */
+  @Override
+  public DefaultNSMessage decode(final byte[] data) {
+
+    try (final ByteArrayInputStream bais = new ByteArrayInputStream(data)) {
+      try (final DataInputStream dais = new DataInputStream(bais)) {
+        final String connectionFactoryId = dais.readUTF();
+        final Identifier srcId = factory.getNewInstance(dais.readUTF());
+        final Identifier destId = factory.getNewInstance(dais.readUTF());
+        final int size = dais.readInt();
+        final List list = new ArrayList(size);
+        final boolean isStreamingCodec = isStreamingCodecMap.get(connectionFactoryId);
+        final Codec codec = connectionFactoryMap.get(connectionFactoryId).getCodec();
+
+        if (isStreamingCodec) {
+          for (int i = 0; i < size; i++) {
+            list.add(((StreamingCodec) codec).decodeFromStream(dais));
+          }
+        } else {
+          for (int i = 0; i < size; i++) {
+            final int byteSize = dais.readInt();
+            final byte[] bytes = new byte[byteSize];
+            dais.read(bytes);
+            list.add(codec.decode(bytes));
+          }
+        }
+
+        return new DefaultNSMessage(
+            connectionFactoryId,
+            srcId,
+            destId,
+            list
+        );
+      }
+    } catch (final IOException e) {
+      throw new RuntimeException("IOException", e);
```

Comment:
There is nothing we can do about it at the moment.

On Sun, Jul 5, 2015 at 11:10 AM, Tae-Geon Um <notifications@github.com>
wrote:

> In
> lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNSMessageCodec.java
> <https://github.com/apache/incubator-reef/pull/255#discussion_r33891386>:
>
> > +            final int byteSize = dais.readInt();
> > +            final byte[] bytes = new byte[byteSize];
> > +            dais.read(bytes);
> > +            list.add(codec.decode(bytes));
> > +          }
> > +        }
> > +
> > +        return new DefaultNSMessage(
> > +            connectionFactoryId,
> > +            srcId,
> > +            destId,
> > +            list
> > +        );
> > +      }
> > +    } catch (final IOException e) {
> > +      throw new RuntimeException("IOException", e);
>
> Can you please explain why it is uncomfortable?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-reef/pull/255/files#r33891386>.
>



-- 
Byung-Gon Chun